Updated Nil-Checking section of the README.

lpollard11 opened the pull request:
Updated the Nil-Checking section of the README with three new cases.
Included standard for the below cases:
-If we want to enter a scope if at least one object is non-nil
-If we want to enter a scope if the object is nil
-If we want to enter a scope if the an object is non-nil/nil and a bool is true/false.

lpollard11 committed:
Updated Nil-Checking section of the README.

Haud commented on README.md:
````diff
@@ -519,6 +519,30 @@ func isError(error: Error?) -> Bool {
+Additionally, check directly for nil if you are checking for nil AND checking if a bool is YES or NO.
+
+```swift
+if item != nil && propertyBool == true {
````
I'm not sure this case is as clear cut as the other two. I think it'd be better as:

`if let _ = nil where propertyBool`

lpollard11 replied:
That's definitely makes sense if you are only checking one propertyBool. How would we use this method if we are checking multiple propertyBools in addition to something that might be nil?

Haud replied:
`if let _ = nil where testOne && testTwo`

Haud replied:
where `testOne`, `testTwo` are `Bool`

lpollard11 replied:
I tested that in playground and I'm getting errors. Plus, that would only work if I want to test if both `testOne` and `testTwo` are `true` or `false`. What if I want to check if `testOne` is `false` and `testTwo` is `true`?

Haud replied:
Can you tell me what errors you are getting. This works for me in playground:

let testOne = true
let testTwo = false

if let str = str where testOne && !testTwo {

}

lpollard11 replied:
It works! Nice, I'll update my pr now

lpollard11 committed:
Edited one nil checking convention

Edited convention for when we want to check for non-nil and bool properties before entering scope.

paulmiard commented:
Feels more like best practices, or even tips to me. Good tips. Yet not sure it should live there

Haud commented:
I'm ok with either approach; Adding more examples for various use cases to the ruleset is perfectly valid IMO

paulmiard commented:
Ok. I'm not against it either. And it makes more sense to have it right after the rule. But there is a difference between rule and examples, and as the guide grows it might get to a point it's too big. We'll see if/when we get there

ThibaultKlein commented on README.md:
````diff
@@ -519,6 +519,30 @@ func isError(error: Error?) -> Bool {
+If there's a case where you want to enter a scope if AT LEAST ONE item is non-nil then use the `??` operator.
+
+```swift
+if let _ = item ?? item1 {
````
is that something Apple is recommending? In Swift 3 for example `--` and `++` are deprecated, just curious

Haud replied:
null-coalescing operator is not going away,

ThibaultKlein replied:
IMHO this shouldn't be a standard, it's a good operator to save some lines but it doesn't provide more clarity to the code (it's even the opposite if you are not familiar with it). We shouldn't force developers to do it.

lpollard11 replied:
I wanted a way to enter a scope if at least one object in a series of objects is not nil using `if let`. In that case should we support just checking directly for nil?

ThibaultKlein replied:
I'm good having that in the best practices

ThibaultKlein commented on README.md:
````diff
@@ -519,6 +519,30 @@ func isError(error: Error?) -> Bool {
+If you want to enter a scope, only if an object is nil, then you can check directly for nil.
+
+```swift
+if item == nil {
````
is that necessary? I think here you are just explaining how to use a feature, not defining a best practice, thoughts?

lpollard11 replied:
Well I wanted to include a case where we don't want to use `if let` to check for nil and its fine to directly check for nil

ThibaultKlein replied:
same comment, i'm good if it's part of the best practices

ThibaultKlein commented:
@lpollard11 would you mind moving this to the `Best Practices` section and create a link from the style guide nil-checking rule?

lpollard11 commented:
Yeah for sure! @ThibaultKlein Do you want me to move all of it or just part of it?

ThibaultKlein commented:
Move it all please :)

Haud commented:
@lpollard11 @ThibaultKlein I'd also maybe add a link in this section referring to it in the best practices so  we can point it out. I think it's valuable information and my worry is that if we don't directly link back to the best practices guide, no one will ever use it / know of it.

paulmiard commented:
That's something we might do for other rules as well, would be good to have a format for it (the same way we have the _rationale_ thing. Give it a try we'll see how it looks

lpollard11 commented:
Sounds good all, I'll move it all to the best practice and include a link to that in this nil-checking section in the actual style guide.

lpollard11 committed:
Removed nil-checking lines from style guide

lpollard11 committed:
Added suggestions for nil checking.

lpollard11 committed:
Added link to best practices.

ThibaultKlein commented on README.md:
````diff
@@ -519,6 +519,8 @@ func isError(error: Error?) -> Bool {
 }
 ```
 
+For style suggestions regarding nil checking visit our [best practices](https://github.com/prolificinteractive/swift-style-guide/blob/master/BestPractices.md) section.
````
`[Best Practices]` I'd assume for better styling

ThibaultKlein commented:
I left a comment, still approving since it's not a huge deal

:+1:

Haud commented:
:+1:
